bot/cogs/extensions.py

```diff
@@ -0,0 +1,231 @@
+        embed = Embed()
+        lines = []
+        cogs = {}
+
+        embed.colour = Colour.blurple()
+        embed.set_author(
+            name="Python Bot (Cogs)",
+            url=URLs.github_bot_repo,
+            icon_url=URLs.bot_avatar
+        )
+
+        for key, _value in self.cogs.items():
+            if "." not in key:
+                continue
+
+            if key in self.bot.extensions:
+                cogs[key] = True
+            else:
+                cogs[key] = False
+
+        for key in self.bot.extensions.keys():
+            if key not in self.cogs:
+                cogs[key] = True
+
+        for cog, loaded in sorted(cogs.items(), key=lambda x: x[0]):
+            if cog in self.cogs:
+                cog = self.cogs[cog]
+
+            if loaded:
+                status = Emojis.status_online
+            else:
+                status = Emojis.status_offline
+
+            lines.append(f"{status}  {cog}")
+
+        log.debug(f"{ctx.author} requested a list of all cogs. Returning a paginated list.")
+        await LinePaginator.paginate(lines, ctx, embed, max_size=300, empty=False)
+
+    async def batch_reload(self, *extensions: str) -> str:
+        """
+        Reload given extensions and return a message with the results.
+
+        If `*` is given, all currently loaded extensions will be reloaded along with any other
+        specified extensions. If `**` is given, all extensions, including unloaded ones, will be
+        reloaded.
+        """
+        failures = {}
+
+        if "**" in extensions:
+            to_reload = EXTENSIONS
+        elif "*" in extensions:
+            to_reload = set(self.bot.extensions.keys()) | set(extensions)
+            to_reload.remove("*")
+        elif extensions:
+            to_reload = extensions
+        else:
+            to_reload = self.bot.extensions.copy().keys()
+
+        for extension in to_reload:
+            _, error = self.manage(extension, Action.RELOAD)
+            if error:
+                failures[extension] = error
+
+        emoji = ":x:" if failures else ":ok_hand:"
+        msg = f"{emoji} {len(to_reload) - len(failures)} / {len(to_reload)} extensions reloaded."
+
+        if failures:
+            failures = '\n'.join(f'{ext}\n    {err}' for ext, err in failures.items())
+            msg += f'\nFailures:```{failures}```'
+
+        log.debug(f'Reloaded all extensions.')
+
+        return msg
+
+    def manage(self, ext: str, action: Action) -> t.Tuple[str, t.Optional[str]]:
+        """Apply an action to an extension and return the status message and any error message."""
+        verb = action.name.lower()
+        error_msg = None
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            if action is Action.RELOAD:
+                # When reloading, just load the extension if it was not loaded.
+                return self.manage(ext, Action.LOAD)
```

Comment:
Is implicitly loading an extension if it isn't loaded undesirable behaviour? If not, another candidate would be to implicitly reload when `!ext load` is done on an already loaded extension.

Reply:
I think that the intention of loading the cog is clear when someone tries to reload a cog, so I don't mind this behavior. The alternative proposal, reloading when someone tries to load something, sounds more counter-intuitive to me. If I recall correctly, the previous version with the misnomer `cogs` denied both, but I often found it annoying that it wouldn't just load an extension on `reload` when it wasn't loaded.

Reply:
Yeah, I felt the same way with the second proposal. Leaving this unresolved for now in case anyone else has opinions to share.

Reply:
I think it's fine, too.